src/Compilers/Core/Portable/InternalUtilities/EnumerableExtensions.cs

```diff
@@ -264,33 +264,48 @@ public static bool IsSingle<T>(this IEnumerable<T> list)
         }
 
         public static bool IsEmpty<T>(this IEnumerable<T> source)
+        {
+            if (source.TryGetCount(out var count))
+            {
+                return count == 0;
+            }
+
+            foreach (var _ in source)
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        public static bool TryGetCount<T>(this IEnumerable<T> source, out int count)
         {
             if (source is IReadOnlyCollection<T> readOnlyCollection)
             {
-                return readOnlyCollection.Count == 0;
+                count = readOnlyCollection.Count;
+                return true;
             }
 
             if (source is ICollection<T> genericCollection)
             {
-                return genericCollection.Count == 0;
+                count = genericCollection.Count;
+                return true;
             }
 
             if (source is ICollection collection)
             {
-                return collection.Count == 0;
+                count = collection.Count;
+                return true;
             }
 
             if (source is string str)
             {
-                return str.Length == 0;
-            }
-
-            foreach (var _ in source)
-            {
-                return false;
+                count = str.Length;
+                return true;
             }
 
-            return true;
+            count = 0;
```

Comment:
Here I would use some value that would generally be invalid for a count. Like ```-1```, for example.

Reply:
Reverted this part of the change